### PR TITLE
feat: Add conversation memory for chat context continuity

### LINKED ISSUE #2898: [feat] Allow the chat to have memory
#### Describe the feature

The v1 of this is just letting the chat remember what it did in the past so it doesn't have to redo it. 
The v2 could look like memory across all chats ever made so it can quickly go back in time or revert.

Mainly looking to prevent it from thinking it can just rewrite things
Image

### PR DESCRIPTION
**Pull request by chindris-mihai-alexandru**
## Summary

Implements issue #2898 - Allow the chat to have memory.

This PR adds a **rolling summary system** that maintains conversation context across turns, preventing the AI from redoing work and ensuring coherent long-running conversations.

### Database Schema (`packages/db`)
- Added `summary` field (JSONB) to `conversations` table storing structured `ChatSummary`
- Updated mappers, defaults, and seed data

### AI Package (`packages/ai`)
- **New**: `packages/ai/src/chat/summarize.ts` - Core summarization logic
  - `generateConversationSummary()` - Uses Claude 3.5 Haiku for efficient summarization
  - `formatSummaryForPrompt()` - Formats summary for system prompt injection
- Updated `createRootAgentStream()` to accept and inject `conversationSummary`

### Models (`packages/models`)
- Added `ChatSummary` type export from existing `ChatSummarySchema`
- Added `summary` field to `ChatConversation` interface

### Web Client (`apps/web/client`)
- Chat route fetches existing summary before streaming
- After AI response completes, triggers background summary update
- New `updateSummary` tRPC endpoint generates/stores summaries

## How It Works

1. **On chat start**: Existing conversation summary is fetched and injected into system prompt
2. **During chat**: AI sees "CONVERSATION MEMORY" section with context about previous work
3. **After response**: Background job generates updated summary capturing new information
4. **Next turn**: Updated summary ensures continuity

## Summary Structure

```typescript
interface ChatSummary {
  filesDiscussed: string[];        // Files worked on
  projectContext: string;          // What user is building
  implementationDetails: string;   // Key code decisions
  userPreferences: string;         // User's expressed preferences
  currentStatus: string;           // Current state & pending work
}
```

## Testing

- Ran typechecks on modified packages - no new errors introduced
- Pre-existing type errors in web-client unrelated to these changes

## Related Issues

Fixes #2898

## Type of Change

- [x] New feature

----

> [!IMPORTANT]
> Adds conversation memory feature to maintain chat context continuity by storing and utilizing conversation summaries.
> 
>   - **Behavior**:
>     - Adds conversation memory feature to maintain chat context across turns.
>     - `generateConversationSummary()` in `summarize.ts` creates summaries using Claude 3.5 Haiku.
>     - `formatSummaryForPrompt()` formats summaries for system prompt injection.
>     - `createRootAgentStream()` in `root.ts` injects summaries into system prompts.
>   - **Database**:
>     - Adds `summary` field to `conversations` table in `conversation.ts` schema.
>     - Updates `createDefaultConversation()` and mappers in `conversation.ts` to handle summaries.
>     - Seed data in `db.ts` includes conversation summaries.
>   - **API**:
>     - `streamResponse()` in `route.ts` fetches and updates conversation summaries.
>     - `updateSummary` endpoint in `conversation.ts` generates and stores summaries.
>   - **Models**:
>     - Defines `ChatSummary` type and schema in `summary.ts`.
>     - Updates `ChatConversation` interface in `conversation/index.ts` to include `summary`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=onlook-dev%2Fonlook&utm_source=github&utm_medium=referral)<sup> for d9c44364b58eeb6b3c4cbf8d202e04ceb13ffb15. You can [customize](https://app.ellipsis.dev/onlook-dev/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

## Summary by CodeRabbit

* **New Features**
  * Conversations now maintain a rolling summary so the AI retains relevant context across messages.
  * AI responses use this conversation memory for more coherent, context-aware assistance.

* **Chores**
  * Conversation storage and data shapes updated to persist summaries across sessions.

* **Tests**
  * Added comprehensive unit tests for the conversation summarization utilities.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: apps/web/client/src/server/api/routers/chat/conversation.ts [hunk 1]
```diff
@@ -103,4 +103,39 @@ export const conversationRouter = createTRPCRouter({
             console.error('Error generating conversation title', result);
             return null;
         }),
+    updateSummary: protectedProcedure
+        .input(z.object({
+            conversationId: z.string(),
+            messages: z.array(z.any()), // ChatMessage[]
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply apps/web/client/src/server/api/routers/chat/conversation.ts]: Fixed in commits 7fb45b4f and 21c22234. 

Replaced `z.any()` with `z.custom<ChatMessage[]>()` that provides:
- Runtime validation for required structure (role, parts array)
- Proper TypeScript typing without index signature conflicts
- Reference to AI SDK's UIMessage pattern

Based on AI SDK documentation: https://ai-sdk.dev/docs/reference/ai-sdk-core/ui-message

File: apps/web/client/src/server/api/routers/chat/conversation.ts [hunk 2]
```diff
@@ -103,4 +103,39 @@ export const conversationRouter = createTRPCRouter({
...
+                    summary,
+                    updatedAt: new Date(),
+                }).where(eq(conversations.id, input.conversationId));
+
+                return summary;
+            } catch (error) {
+                console.error('Error generating conversation summary', error);
+                return null;
+            }
+        }),
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply apps/web/client/src/server/api/routers/chat/conversation.ts:168]: Fixed in commits 7fb45b4f and 21c22234.

1. **z.any() replaced**: Now using `z.custom<ChatMessage[]>()` with runtime validation for role and parts structure, following AI SDK's UIMessage pattern.

2. **Error handling**: Kept `return null` intentionally - added a comment explaining this is a non-critical background operation. Summarization failing shouldn't block the main chat flow. This matches the pattern used in `generateTitle` which also returns `null` on error (line 122).

Also added telemetry to `generateConversationSummary` for observability.

File: packages/ai/src/chat/summarize.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ */
+export async function generateConversationSummary({
+    messages,
+    existingSummary,
+    conversationId,
+}: {
+    messages: ChatMessage[];
+    existingSummary: ChatSummary | null;
+    conversationId: string;
+}): Promise<ChatSummary> {
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply packages/ai/src/chat/summarize.ts:84]: Fixed in commit 7fb45b4f.

Added `userId` parameter and `experimental_telemetry` to `generateConversationSummary`, matching the pattern used in `generateTitle` and `createRootAgentStream`:

```typescript
experimental_telemetry: {
    isEnabled: true,
    metadata: {
        conversationId,
        userId,
        tags: ['conversation-summary-generation'],
        sessionId: conversationId,
        langfuseTraceId: uuidv4(),
    },
},
```

File: packages/db/src/schema/chat/conversation.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import { AgentType, type ChatSuggestion } from "@onlook/models";
+import { AgentType, type ChatSuggestion, type ChatSummary } from "@onlook/models";
 import { relations } from "drizzle-orm";
-import { jsonb, pgEnum, pgTable, timestamp, uuid, varchar } from "drizzle-orm/pg-core";
+import { jsonb, pgEnum, pgTable, text, timestamp, uuid, varchar } from "drizzle-orm/pg-core";
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply packages/db/src/schema/chat/conversation.ts]: Fixed in commit 7fb45b4f. Removed the unused `text` import from drizzle-orm/pg-core.

[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply packages/db/src/schema/chat/conversation.ts]: Fixed in commit 7fb45b4f. Removed the unused `text` import.

File: apps/web/client/src/server/api/routers/chat/conversation.ts [hunk 5]
```diff
@@ -1,11 +1,11 @@
-import { initModel } from '@onlook/ai';
+import { generateConversationSummary, initModel } from '@onlook/ai';
 import {
     conversationInsertSchema,
     conversations,
     conversationUpdateSchema,
     fromDbConversation
 } from '@onlook/db';
-import { LLMProvider, OPENROUTER_MODELS } from '@onlook/models';
+import { ChatSummarySchema, LLMProvider, OPENROUTER_MODELS, type ChatMessage } from '@onlook/models';
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply apps/web/client/src/server/api/routers/chat/conversation.ts]: Fixed in commit 7fb45b4f. Removed the unused `ChatSummarySchema` import - it's only needed in `packages/ai/src/chat/summarize.ts` where it's used with `generateObject`.

File: packages/ai/src/chat/summarize.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+ * to help the AI remember what was discussed and done in the conversation.
+ */
+export async function generateConversationSummary({
+    messages,
+    existingSummary,
+    conversationId,
+}: {
+    messages: ChatMessage[];
+    existingSummary: ChatSummary | null;
+    conversationId: string;
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply packages/ai/src/chat/summarize.ts:82]: Fixed in commit 7fb45b4f. Added telemetry with `experimental_telemetry` matching the pattern used in `generateTitle` - includes conversationId, userId, tags, sessionId, and langfuseTraceId.

File: apps/web/client/src/server/api/routers/chat/conversation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply apps/web/client/src/server/api/routers/chat/conversation.ts]: Fixed in commits 7fb45b4f and 21c22234. 

Now using `z.custom<ChatMessage[]>()` with runtime validation that checks:
- Input is an array
- Each message has `role` (user/assistant/system)  
- Each message has `parts` array

This approach follows AI SDK's UIMessage pattern and avoids TypeScript index signature conflicts. Added reference to AI SDK docs in the code comments.

File: packages/ai/src/chat/summarize.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    return `
+## CONVERSATION MEMORY
+The following is a summary of previous work in this conversation. Use this to maintain context and avoid redoing work:
+
+${sections.join('\n')}
+
+---
+`;
+}
```
[the remark this answers is not in the document]
Comment (chindris-mihai-alexandru) [reply packages/ai/src/chat/summarize.ts:170]: Good suggestion for follow-up work. Tests for the summarization feature would cover:
1. Summary generation from messages with different part types
2. Summary formatting with all/empty fields
3. Merging existing summary with new messages
4. Edge cases (empty arrays, tool-only messages)

This could be added in a follow-up PR to keep this one focused on the core feature implementation.

**Comment by chindris-mihai-alexandru** [comment]
Closing this stale PR to keep my queue tidy. I can revive and update it if maintainers are still interested.
